torchdynamo/optimizations/distributed.py

```diff
@@ -0,0 +1,105 @@
+from typing import Dict
+from typing import List
+
+import torch
+import torch.fx as fx
+
+
+class DDPOptimizer:
+    def __init__(
+        self, bucket_bytes_cap: int, parameters_to_ignore: List[str], backend_compile_fn
+    ):
+        self.bucket_bytes_cap = bucket_bytes_cap
+        self.parameters_to_ignore = parameters_to_ignore
+        self.backend_compile_fn = backend_compile_fn
+
+    def compile_fn(self, gm: fx.GraphModule, example_inputs: List[torch.Tensor]):
+        """
+        TODO:
+        - handle params_and_buffers_to_ignore
+        - implement call_method/call_forward
+        - handle kwargs
+        """
+
+        # compute the partition map according to DDP bucket logic
+        bucket_bytes = 0
+        bucket_actual_sizes = []
+        node_splits = [[]]
+        for node in reversed(gm.graph.nodes):
+            if bucket_bytes >= self.bucket_bytes_cap:
+                bucket_actual_sizes.insert(0, bucket_bytes)
+                bucket_bytes = 0
+                node_splits.append([])
+
+            if node.op == "output" or node.op == "placeholder":
+                continue
+
+            elif node.op == "call_module":
+                target = gm.get_submodule(node.target)
```

Comment:
I believe they are guaranteed to be modules at the same level of granularity to each other, since they are produced by the partitioner pass.  In other words, they should be non-leaf modules and they should contain other modules that could be non-leaf or leaf modules.